Backports release 1.11

KristofferC opened the pull request:
Backported PRs:
- [x] #53757 <!-- Add an IndexStyle example to the diagind docstring -->
- [x] #53809 <!-- Add missing GC_POP() in emit_cfunction -->
- [x] #53789 <!-- also check that UUID of project is non-null when treating it as a package -->
- [x] #53805 <!-- precompilepkgs: simplify custom config printing if only one -->
- [x] #53822 <!-- Bump libuv -->
- [x] #53837 <!-- update MPFR to 4.2.1 -->
- [x] #53862 <!-- precompilepkgs: fix error reporting -->
- [x] #53774 <!-- Remove some duplicates from emitted compilation traces -->
- [ ] #53696 <!-- add invokelatest to on_done callback in bracketed paste -->
- [x] #53383 <!-- Add `_unsetindex!` methods for `SubArray`s and `CartesianIndex`es -->
- [x] #53475 <!-- Fix boundscheck in unsetindex for SubArrays -->
- [x] #53888 
- [x] #53870 <!-- Revert change to checksum for llvm-julia -->
- [x] #53906 <!-- Add `Base.isrelocatable(pkg)` -->
- [x] #53833 <!-- Profile: make heap snapshots viewable in vscode viewer -->
- [x] #53961 <!-- `LazyString` in `LinearAlgebra.checksquare` error message -->
- [x] #53962 <!-- Use StringMemory instead of StringVector where possible -->
- [x] #53825 <!-- profile: doc: update the `Allocs.@profile` doc string -->
- [x] #53975 <!-- `LazyString` in `DimensionMismatch` error messages in broadcasting -->
- [x] #53905 <!-- Avoid repeated precompilation when loading from non-relocatable cachefiles -->
- [x] #53896 <!-- Make reshape and view on Memory produce Arrays and delete wrap -->
- [x] #53991 <!-- Test and fix non-int-length bug in `view(::Memory, ::Union{UnitRange, Base.OneTo})` -->

Non-merged PRs with backport label:
- [ ] #53976 <!-- LinearAlgebra: LazyString in interpolated error messages -->
- [ ] #53957 <!-- tweak how filtering is done for what packages should be precompiled -->
- [ ] #53943 <!-- Allow for querying of build_id from objects -->
- [ ] #53941 <!-- Fix writing of AnnotatedChars to AnnotatedIOBuffer -->
- [ ] #53882 <!-- Warn about cycles in extension precompilation -->
- [ ] #53801 <!-- Normalise annotations on AnnotatedIOBuffer read -->
- [ ] #53800 <!-- Ensure that annotations are well-ordered in optimisation -->
- [ ] #53794 <!-- Tweak annotation ordering during modifications -->
- [ ] #53750 <!-- inference correctness: fields and globals can revert to undef -->
- [ ] #53715 <!-- Add read/write specialisation for IOContext{AnnIO} -->
- [ ] #53707 <!-- Make ScopedValue public -->
- [ ] #53660 <!-- put Logging back in default sysimage -->
- [ ] #53509 <!-- revert moving "creating packages" from Pkg.jl -->
- [ ] #53452 <!-- RFC: allow Tuple{Union{}}, returning Union{} -->
- [ ] #53402 <!-- Add `jl_getaffinity` and `jl_setaffinity` -->
- [ ] #52694 <!-- Reinstate similar for AbstractQ for backward compatibility -->
- [ ] #51928 <!-- Styled markdown, with a few tweaks -->
- [ ] #51811 <!-- Make banner size depend on terminal size -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->

jishnub committed:
Add `_unsetindex!` methods for `SubArray`s and `CartesianIndex`es (#53383)

With this, the following (and equivalent calls) work:
```julia
julia> copyto!(view(zeros(BigInt, 2), 1:2), Vector{BigInt}(undef,2))
2-element view(::Vector{BigInt}, 1:2) with eltype BigInt:
 #undef
 #undef

julia> copyto!(view(zeros(BigInt, 2), 1:2), view(Vector{BigInt}(undef,2), 1:2))
2-element view(::Vector{BigInt}, 1:2) with eltype BigInt:
 #undef
 #undef
```

Close https://github.com/JuliaLang/julia/issues/53098. With this, all
the `_unsetindex!` branches in `copyto_unaliased!` work for
`Array`-views, and this makes certain indexing operations vectorize and
speed-up:
```julia
julia> using BenchmarkTools

julia> a = view(rand(100,100), 1:100, 1:100); b = view(similar(a), axes(a)...);

julia> @btime copyto!($b, $a);
  16.427 μs (0 allocations: 0 bytes) # master
  2.308 μs (0 allocations: 0 bytes) # PR
```

Improves (but doesn't resolve)
https://github.com/JuliaLang/julia/issues/40962 and
https://github.com/JuliaLang/julia/issues/53158

```julia
julia> a = rand(40,40); b = rand(40,40);

julia> @btime $a[1:end,1:end] .= $b;
  5.383 μs (0 allocations: 0 bytes) # v"1.12.0-DEV.16"
  3.194 μs (0 allocations: 0 bytes) # PR
```
ƒ
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

(cherry picked from commit 1a9040908b9bb89927b2d318da3f9fe1e457abea)

jishnub committed:
Fix boundscheck in unsetindex for SubArrays (#53475)

These had been copy-pasted incorrectly, and should throw an error if the
indices are out of bounds.

(cherry picked from commit 98b3f721275082cbf3c1ece9d3ae394ec305e12f)

jishnub committed:
Add an IndexStyle example to the diagind docstring (#53757)

Also, simplifies the docstring by splitting the usage into two lines.

(cherry picked from commit 0b1587c958583030bc3a61114858f04fc6802bfd)

fatteneder committed:
remove dllexport `jl_arraylen` which no longer exists (#53765)

(cherry picked from commit 1d532f4fc798ecaa5bcacfcedc4c36143ad0e68f)

kpamnany committed:
Remove some duplicates from emitted compilation traces for Julia 1.10 (#53776)

When multiple threads concurrently attempt to compile the same method,
`--trace-compile` could emit duplicate `precompile` statements. This
small tweak eliminates one source of these duplicates.

Backported to 1.10 from https://github.com/JuliaLang/julia/pull/53774.

(cherry picked from commit f6a3c5f8d07faea45653b34a74175d7e8a46999b)

gbaraldi committed:
Add missing GC_POP() in emit_cfunction (#53809)

~~Apparently somewhere in codegen inside `emit_codeinst`some piece of
code is relying on the implicit try catch gcstack restoring. I haven't
got the analyzer working on that file yet (it has hundreds of issues and
it doesn't like C++ code that much + the file is tens of thousands of
lines after includes so it struggles).~~

This fixes the compileall segfault in apple-aarch64 ci.
https://github.com/JuliaLang/julia/issues/53811

(cherry picked from commit 52fc79696e0b08fe1793a001daaefd1d91d59f0c)

KristofferC committed:
also check that UUID of project is non-null when treating it as a package (#53789)

Fixes https://github.com/JuliaLang/julia/issues/53788

(cherry picked from commit 3e37e175d3c2c18de6927db7a4f5ae1d446f6b23)

IanButterworth committed:
precompilepkgs: simplify custom config printing if only one (#53805)

Currently it's a bit excessive in the `Pkg.test` precompile job

![Screenshot 2024-03-20 at 12 04
03 PM](https://github.com/JuliaLang/julia/assets/1694067/7600f0b8-6e4b-43b2-9c42-c8d5d16b8d57)

This PR
```
Precompiling project for configuration --code-coverage=none --color=yes --check-bounds=yes --warn-overwrite=yes --depwarn=yes --inline=yes --startup-file=no --track-allocation=none...
    354.9 ms  ✓ RFFT
  1 dependency successfully precompiled in 1 seconds. 38 already precompiled.
```

Pkg could also just set the non-default flags to minimize the list.

(cherry picked from commit 92918457e31e3696fff9fa3ea91aaa5e8159a23a)

Keno committed:
Bump libuv (#53822)

Fixes https://github.com/JuliaDebug/Cthulhu.jl/issues/541 by picking up
https://github.com/libuv/libuv/commit/3ecce914105590e6112cd0c6d4f6b30ac6a6c24f.

(cherry picked from commit 5ed51d33c50f665200e562fe0cfb8ff70e566d59)

fxcoudert committed:
update MPFR to 4.2.1 (#53837)

MPFR 4.2.1 was released on 22 August 2023.
It contains bugfixes.

(cherry picked from commit 61caaa886afe18273e2f7b7d09a714298a62a9b3)

IanButterworth committed:
precompilepkgs: fix error reporting (#53862)

(cherry picked from commit fc2d3af5bd8b6a0ba8d057c31129e9318b19f2c5)

JeffBezanson committed:
add invokelatest to on_done callback in bracketed paste (#53696)

fixes #52120

(cherry picked from commit bb3b09de50720d4b47c34fd53638e292c50c695e)

KristofferC commented:
@nanosoldier `runtests(["UsingMerge", "ScoredTests", "TextWrap", "StructIO", "MetadataStreams", "PrecompileMacro", "Undefs", "Exceptions", "CachedFunctions", "NumericIO", "LLVMLoopInfo", "LicenseGrabber", "DynamicallyLoadedEmbedding", "ConcreteTupleDicts", "EnumX", "PropertyDicts", "ProtoStructs", "BitFlags", "ExternalDocstrings", "WidthLimitedIO", "DataBags", "TwoDimensional", "PairedLinkedLists", "Expect", "Ramps", "Elliptic", "BayesHistogram", "FlagSets", "HierarchicalMatrices", "Why", "ClimateEasy", "TestExtras", "TrueSkillThroughTime", "AutoHashEquals", "MetaUtils", "ScopedValues", "Microgrids", "CubicSplines", "TypeUtils", "TestItemDetection", "SentinelViews", "ParallelProcessingTools", "Lebedev", "StructTypes", "StableLinearAlgebra", "Kirei", "Checkerboard", "NeighborJoining", "Bobby", "DynamicObjects", "Memoize", "StaticNumbers", "SimpleIntegrals", "CompositeStructs", "SaferIntegers", "DiracNotation", "Bessels", "Preferences", "RecursiveDicts", "Chairmarks", "Emojify", "ConstructionBase", "NormalHermiteSplines", "DateFormats", "Skipper", "RandomNumbers", "LogSpline", "StructuredArrays", "Maxvol", "ROCCurves", "DataPipes", "LazyGrids", "JuliaWorkspaces", "Hygienic", "RegularizedCovarianceMatrices", "CompileTraces", "SPGBox", "EnsembleInversion", "SwapStreams", "NativeNaNMath", "LoopFieldCalc", "YoungTableaux", "DynamicSparseArrays", "Match", "XAIBase", "SubspacePreservingSparsification", "EasyJobsBase", "InvertedIndices", "PyFormattedStrings", "TypedSyntax", "LoweredCodeUtils", "SIMDscan", "Resizing", "Umlaut", "PlutoVSCodeDebugger", "PlutoHooks", "StructEquality", "PreferenceTools", "DictTools", "StaticStrings", "RegistryTools", "StandardPacked", "PDMats", "ProperOrthogonalDecomposition", "ScanByte", "IterationControl", "SqraCore", "TestItemRunner", "ChunkSplitters", "AhoCorasickAutomatons", "ExtendedKronigPennyMatrix", "IntensityMetrics", "UInt12Arrays", "OpenTelemetryAPI", "LibLZO", "GZip", "TopoChains", "Unitful", "NamedColors", "ZipFile", "TerminalRegressionTests", "JSONRPC", "ConvolutionOperators", "Aqua", "QuadGK", "FastRunningMedian", "AesKeywrapNettle", "Inertia", "QMDDocTools", "Reactive", "RootsAndPoles", "MicroMamba", "DecisionTree", "ProtoBuf", "DataStructures", "ConstraintCommons", "Classes", "PropCheck", "TableShowUtils", "CodeTools", "SequentialMonteCarlo", "CodecBGZF", "UnitTypes", "MaybeInplace", "REPLference", "ColPack", "UnionCollections", "RetroCap", "Darknet", "DocumentationOverview", "ObjectOriented", "LocalRegistry", "FillArrays", "Accessors", "MuJoCo", "HalfEdges", "UProbes", "LLVMCGUtils", "CloseOpenIntervals", "StaticArrayInterface", "PDMatsExtras", "SourceWalk", "InteractiveErrors", "IntervalSets", "LLVM", "VimBindings", "BSDiff", "DiffEqDiffTools", "NodeCall", "OMEinsumContractionOrders", "CellArrays", "XUnit", "Countries", "LACosmic", "CloughTocher2DInterpolation", "Revise", "HomogeneousTransformations", "LasIO", "MaxPlus", "Krylov", "PosteriorDB", "ArrayInterfaceBlockBandedMatrices", "IntervalUnions", "NamedArrays", "RootedTrees", "ManifoldsBase", "KissSmoothing", "StationXML", "MCPhyloTree", "BlockDiagonals", "GPUCompiler", "VariantCallFormat", "StableHashTraits", "Comonicon", "EvolutionaryModelingTools", "CondaPkg", "PandasJSON", "HypertextTemplates", "SemiseparableMatrices", "MutableArithmetics", "AllocCheck", "LogDensityProblemsAD", "SimpleWorkflows", "Polylogarithms", "SimpleConfig", "Caching", "SQLiteDF", "ImageCore", "Metaheuristics", "MolecularMinimumDistances", "PSIS", "CategoricalArrays", "SparsityTracing", "WriteVTK", "CamiFITS", "Distances", "FunctionProperties", "GNSSDecoder", "QuasiArrays", "Rotations", "LinearFold", "InterferometricModels", "ThreadedSparseCSR", "Ewalder", "FunnyORM", "DataInterpolations", "LongDatasetSort", "SparseIR", "CellListMap", "Compose", "ImageShow", "CartesianJoin", "BigWig", "Merly", "GoogleSheetsCSVExporter", "Metis", "BandedMatrices", "PlutoVista", "HTTP", "BertWordPieceTokenizer", "RecursiveArrayTools", "KernelAbstractions", "BasicBSpline", "ManifoldDiff", "Interpolations", "FixedWidthTables", "RelationalDatasets", "DLMReader", "Electrum", "Peridynamics", "ProbabilisticRoadMap", "Graphs", "LinearMaps", "DirectSearch", "AprilTags", "AsterReader", "OMEinsum", "QEDbase", "DynamicExpressions", "DataDeps", "FiniteDiff", "Colab", "HDF5", "LokiLogger", "ExcelReaders", "DynamicPolynomials", "BlockBandedMatrices", "DictArrays", "BSeries", "DirectionalStatistics", "ServerMetrics", "DTALib", "LanguageServer", "CDCLsat", "HTTPDownloads", "AppliGeneralLedger", "DateSelectors", "AutoBZCore", "MonteCarloTesting", "Skans", "ExaModels", "PeriodicSchurDecompositions", "Mangal", "WebAssemblyCompiler", "MultivariateStats", "VoronoiCells", "DiffusionMap", "DensityRatioEstimation", "GraphPlot", "PartitionedStructures", "DataManipulation", "MAT", "MultiAgentPathFinding", "PlutoTables", "TimeStruct", "PlutoUIExtra", "GCPDecompositions", "TransformVariables", "ImageMagick", "TypedPolynomials", "QuadraticHamiltonians", "BigBed", "FastTransformsForwardDiff", "RAFF", "PDBTools", "FLoops", "CCBlade", "ReadStatTables", "PatternFolds", "VisionHeatmaps", "ImplicitDifferentiation", "Feather", "ADIOS2", "RectiGrids", "TwoFAST", "InMemoryDatasets", "BioCCP", "NetCDF", "GenomicAnnotations", "GPUArrays", "WeibullParetoDist", "InferenceObjects", "Transducers", "ImplicitGlobalGrid", "FactorRotations", "LazyBandedMatrices", "Manopt", "StandardizedPredictors", "MondrianForests", "NLPModelsTest", "ProteinSecondaryStructures", "JunctionTrees", "CloudStore", "VOTables", "MultiPrecisionArrays", "Ghost", "MetalKernels", "SlackThreads", "SeparableFunctions", "ElectroPhysiology", "SQLStore", "MathTeXEngine", "MLJScikitLearnInterface", "TrajectoryOptimization", "SpineBasedRecordLinkage", "WilliamsonTransforms", "InfiniteArrays", "OutlierDetectionPython", "GaussianRandomFields", "HypothesisTests", "SingularSpectrumAnalysis", "Materials", "Vega", "HiddenMarkovModels", "PressureSensitiveMats", "Checkpointing", "NISTStatisticalReferenceData", "AnovaGLM", "Subspaces", "GeneralizedPhase", "Pitaya", "FMIImport", "NaNStatistics", "TensorBoardLogger", "FastDifferentiation", "VegaLite", "OptimizationMetaheuristics", "Intervals", "Legolas", "QuantumControl", "RecursiveFactorization", "AlgebraOfNNs", "SamplingRB", "Chevie", "Shapefile", "CVaRRiskParity", "KomaMRIBase", "ConstraintDomains", "StaticTools", "RData", "ODEConvergenceTester", "IERSConventions", "DatagenCopulaBased", "PlutoStyles", "CaNNOLeS", "AbstractGPs", "GroupedTransforms", "RestrictedBoltzmannMachines", "SunAsAStar", "Onda", "NaiveNASlib", "LegolasFlux", "GraphRecipes", "ContinuumArrays", "Thermochron", "GaussianMixtureRegressions", "Acquisition", "QUBOConstraints", "PlantRayTracer", "Distributions", "ClimaAnalysis", "RedBlackGraph", "NCRegressionTests", "JupyterPlutoConverter", "Arrow", "TransitionMatrices", "ImageNoise", "LightOSM", "KomaMRIFiles", "PseudoPotentialIO", "Groebner", "TimeDag", "FixedEffectModels", "Layered", "JuMP", "DistanceTransforms", "Copulas", "COPT", "ForecastData", "HiGHS", "GaussianMixtures", "ANOVAapprox", "MosekTools", "Gen", "Quantikz", "TuringCallbacks", "ProximalAlgorithms", "EasyFit", "STREAMBenchmark", "AcousticAnalogies", "VirtualObservatory", "ChowLiuTrees", "ClinicalTrialUtilities", "CUDAKernels", "SmoQyDQMC", "IndexSpaces", "ChaChaCiphers", "COSMO", "Clarabel", "CompositionalNetworks", "Trading", "CoulombIntegrals", "StaticCompiler", "QuasinormalModes", "CoinbasePro", "Meshes", "BioFetch", "AutoMLPipeline", "PlmDCA", "NormalizingFlows", "OptimizationPRIMA", "UnitCommitment", "DiffOpt", "NLPModelsJuMP", "NeuralNetworkReachability", "Robotlib", "SkyDomes", "AdvancedMH", "StochasticAD", "AlphaStableDistributions", "QUBOTools", "Tulip", "NaiveBayes", "ImageMorphology", "MatrixEnsembles", "StateSpaceInference", "ComplexityMeasures", "ExplainableAI", "ExtremeStats", "Catlab", "HierarchicalTemporalMemory", "NURBS", "PointEstimateMethod", "MEstimation", "MeasureTheory", "SampleChains", "PlanktonIndividuals", "Optim", "SymArrays", "TensorCast", "MLJLinearModels", "Constraints", "HurdleDMR", "HarmonicOrthogonalPolynomials", "ConstraintSolver", "PowerModelsDistribution", "TSFrames", "TemporalGPs", "CalibrationErrors", "Parquet2", "Polyhedra", "ExaPF", "GroupedErrors", "PolyJuMP", "GaussianBasis", "MPSGE", "SingularIntegrals", "UKCarbonIntensityData", "ParticleCorrelations", "AnnuliOrthogonalPolynomials", "SemiclassicalOrthogonalPolynomials", "ChaoticDynamicalSystemLibrary", "DyadicKDE", "EquilibratedFlux", "HOODESolver", "GraphSignals", "PiecewiseOrthogonalPolynomials", "ODEInterfaceDiffEq", "MLJOpenML", "QMDP", "NoncommutativeGraphs", "PowerModelsAnalytics", "StatisticalGraphics", "ThermodynamicIntegration", "SeisBase", "SSIMLoss", "ApproximateVI", "MolSimToolkit", "GtkReactive", "DroneSurveillance", "Packmol", "OptimalTransport", "LSODA", "WiSER", "MultivariateOrthogonalPolynomials", "SimSpread", "TracyWidomBeta", "MaximinOPF", "GLFixedEffectModels", "ExaAdmm", "ODE", "StLifeIns", "OceanWaveSpectralFitting", "RadialPiecewisePolynomials", "RockSample", "ReplicateBE", "BulkLMM", "Sundials", "ModeCouplingTheory", "TableTransforms", "ClassicalOrthogonalPolynomials", "Winston", "Tabulae", "WhittleLikelihoodInference", "NoiseRobustDifferentiation", "AlgebraicCurveOrthogonalPolynomials", "BasicBSplineExporter", "TestParticle", "QuantumStateBase", "EvidentialFlux", "KeplerGL", "GeoParquet", "ReferenceFiniteElements", "SymbolicRegression", "NonconvexMetaheuristics", "EnsembleKalmanProcesses", "PlotSeis", "FluxPrune", "ApproxFunSingularities", "PlotMesh", "FluxExtra", "LiteQTL", "AlgebraicPetri", "PhyloNetworks", "MPIFiles", "SparseKmeansFeatureRanking", "EconomicScenarioGenerators", "JOLI", "AugmentedGPLikelihoods", "PosteriorStats", "GADM", "DCEMRI", "QuantumStateDistributions", "EDM4hep", "IncrementalPruning", "BridgeDiffEq", "CherenkovDeconvolution", "SpinGlassPEPS", "TransitionsInTimeseries", "WordCloud", "VLBIImagePriors", "NestedSamplers", "BVHFiles", "ARCHModels", "TimeseriesTools", "PlotRNA", "RelevancePropagation", "OpenADMIXTURE", "PLaplace", "GreedyAlign", "EvoTrees", "MCMCTempering", "QXContexts", "KomaMRIPlots", "JetPack", "PlutoStaticHTML", "MITgcmTools", "SignedDistanceFunction", "DynACof", "Alexya", "TimeSeriesClassification", "TeneT", "Variography", "GeoStatsFunctions", "PoreMatMod", "FastaLoader", "Yields", "NablaNet", "PlutoPages", "KnetNLPModels", "Knet", "MLJBalancing", "Braket", "GlobalSensitivity", "QuantumInformation", "SFrontiers", "ElasticSurfaceEmbedding", "Surrogates", "ChemistryFeaturization", "SetIntersectionProjection", "TransformerBlocks", "PottsGumbelRBMLayers", "SMLMBoxer", "QuantumOpticsBase", "AztecDiamonds", "Swalbe", "GumbelSoftmax", "KomaMRICore", "YAXArrays", "IPMeasures", "InferOpt", "GraphNetCore", "QuantitativeSusceptibilityMappingTGV", "PortfolioOpt", "DistributedStwdLDA", "DeepUnfoldedCDLMotif", "Wflow", "SimDNA", "DistributedFlux", "UNet", "Tsunami", "FinanceModels", "MetidaCu", "Hyperelastics", "BackgroundSubtraction", "Hopfields", "TrajOptPlots", "REoptLite", "SurrogatesFlux", "ZonalFlow", "SchwarzChristoffel", "OperatorLearning", "Korg", "IRKGaussLegendre", "DynamicalSystemsBase", "SampleChainsDynamicHMC", "ParallelStencil", "Imbalance", "MIRT", "GeoStatsProcesses", "Mehrotra", "PALEOaqchem", "PALEOcopse", "LLMTextAnalysis", "GeometryOptimization", "BEAST", "ActuaryUtilities", "MultiStateSystems", "GeneralizedSDistributions", "MimiMooreEtAlAgricultureImpacts", "Symbolics", "NBodySimulator", "HierarchicalEOM", "SoleData", "KnetLayers", "ReinforcementLearningCore", "EFTfitter", "EMpht", "UltraDark", "DistributedFactorGraphs", "ImageQuilting", "MriResearchTools", "LaplaceRedux", "MatrixProductBP", "Flux3D", "GeoStatsTransforms", "KiteModels", "DelayDiffEq", "ReachabilityAnalysis", "ConstrainedSystems", "LITS", "NonlinearSchrodinger", "MatrixMerge", "VisualizeMotifs", "FractionalSystems", "DifferentiableTrajectoryOptimization", "CDLmotif", "PartiallySeparableSolvers", "Grumps", "IndividualDisplacements", "RAPIDS", "ReinforcementLearning", "DINCAE", "Consensus", "GenericTensorNetworks", "FSimBase", "PotentialFlow", "BAT", "ColorSchemeTools", "EffectiveWaves", "MOTIFs", "SystemBenchmark", "PlantBiophysics", "HydroPowerSimulations", "ConstraintLearning", "ImageUtils", "EqualitySampler", "ObjectDetector", "PositionVelocityTime", "Fronts", "TrajGWAS", "Mikrubi", "Juice", "OrdinalGWAS", "GeometricProblems", "Futbolista", "PlotPlants", "Plots", "AllenNeuropixelsBase", "SeisPDF", "HighDimPDE", "TensorInference", "ITensorUnicodePlots", "ControlSystemIdentification", "AcousticRayTracers", "TMLE", "RomeoDFT", "DifferentiableStateSpaceModels", "LowRankIntegrators", "ActionModels", "ProcessBasedModelling", "SurfaceReactions", "SideKicks", "Knockoffs", "LocalAnisotropies", "AlgebraicInference", "SurfaceCoverage", "StirredReactor", "Pesto", "QXTools", "PixelArt", "PlugFlowReactor", "BatchReactor", "SurrogatesMOE", "QuantumAnnealing", "JointSurvivalModels", "CalibrateEmulateSample", "Tapestree", "QSimulator", "ConformalPrediction", "BaseModelica", "ScatteringOptics", "NeuronBuilder", "ODEProblemLibrary", "NEOs", "NestedEnvironments", "PolynomialGTM", "ParameterizedFunctions", "NetworkJumpProcesses", "DiffEqProblemLibrary", "NonconvexBayesian", "JumpProblemLibrary", "LinearNoiseApproximation", "BlockSystems", "SDEProblemLibrary", "YaoQX", "AdditiveCellCom", "GalacticPotentials", "IntervalConstraintProgramming", "StateSpaceReconstruction", "ReactionNetworkImporters", "ProbabilisticEchoInversion", "OpenQuantumSystems", "CellMLToolkit", "CryoGrid", "AstrodynamicalModels", "AstrodynamicalSolvers", "MinimallyDisruptiveCurves", "LatticeQCD", "GeneralizedSasakiNakamura", "EHTImages", "StochasticIntegrals", "ModelOrderReduction", "Modia", "GNSSReceiver", "Catalyst", "Bactos", "ValidatedNumerics", "DiffRaster2D", "CorrelationTrackers", "GPDiffEq", "FaultTolerantControl", "DynamicalSystems", "AstroChemistry", "ParticleMDI", "CorrelationFunctions", "TrillionDollarWords", "AtomGraphs", "GraphNets", "BarBay", "TKTDsimulations", "OptimizationMOI", "GeometricFlux", "BinomialSynapses", "SBMLToolkit", "PPLM", "MaterialReconstruction", "FMISensitivity", "ReactionSensitivity", "PQBaseCamp", "PANDA", "MakieRichText", "SymbolicNumericIntegration", "DickeModel", "ClimateTools", "AlgebraicDynamics", "Chron", "GeometricMachineLearning", "EditorsRepo", "SubsidenceChron", "FatDatasets", "AtmosphericDeposition", "CellularPotts", "TuringGLM", "GraphMakie", "CirculatorySystemModels", "CollectiveSpins", "SMLMFrameConnection", "SemanticTrajectories", "Collide", "SequentialSamplingModels", "MakiePDF", "FlowAtlas", "RMLImaging", "GeoIO", "ClimatePlots", "QuantumDynamics", "HmtArchive", "NuclearToolkit", "Vahana", "Cropbox", "WaveletsExt", "StructuredLight", "TSML", "GeoArtifacts", "DataDrivenSR", "QuantumOptics", "SDFResults", "HiQGA", "MINDFulMakie", "Population", "MDToolbox", "ParameterEstimocean", "Biofilm", "NeuralGraphPDE", "FastTabular", "Bagyo", "GeneticsMakie", "Survey", "PhyloClustering", "EnergyCommunity", "CDGRNs", "AlphaZero", "MRIsim", "BloqadeGates", "FastAI", "MiseEnPage", "MixedModelsMakie", "QuanEstimation", "PowerGraphics", "BloqadeODE", "SphereUDE", "DataDrivenLux", "Sisyphus", "NeXLSpectrum", "LighthouseFlux", "FMIFlux", "StochasticGene", "UnfoldCDL", "KiteSimulators", "BifurcationInference", "CEEDesigns", "Turkie", "REPL", "MCPTrajectoryGameSolver", "QuantumAnnealingAnalytics", "MRINavigator", "ReducedBasisMethods", "SMLMSim", "HarmonicBalance", "Chamber", "RigorousInvariantMeasures", "AdaptiveTestPlots", "Causal", "TaijaPlotting", "IterativeLearningControl", "OVERT", "RetentionParameterEstimator", "NeuroAnalysis", "RDMREopt", "RelativisticDynamics", "QuantumCollocation", "TumorGrowth", "VlasovMethods", "GIRFReco", "BondGraphs", "GasChromatographySimulator", "Controlz", "UnfoldMakie", "GasChromatographySystems", "SmartML", "QuantumSavory", "MagNav", "NTNk", "SparseArrays", "ContinuousTimePolicyGradients"], vs = ":release-1.10")`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/53790#issuecomment-2023112098) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c356e60_vs_bd47eca/report.html) is available.

aviatesk committed:
fix typos in codegen.cpp (#53888)

Not sure why this typo didn't mess up the CI, but it looks like a clear
problem, let's correct it. I'd appreciate any idea on how to exercise
this change.

- fixes JuliaDebug/JuliaInterpreter.jl#621

Zentrik committed:
Revert change to checksum for llvm-julia (#53870)

Should fix #53399. I think this is the correct fix.
Makes it so you can build Julia with `USE_BINARYBUILDER=0`.

(cherry picked from commit b2e8eb2253c4641c058f61f3bd5cb3447484b7ed)

fatteneder committed:
Add `Base.isrelocatable(pkg)` (#53906)

This PR adds a utility function `isrelocatable(pkg)` that can be used to
check if `pkg` is already precompiled and if the associated cachefile is
relocatable.

The reason to implicitly perform the `isprecompiled` check is that the
exact same computation needs to be done to find the right `.ji`.

A `pkg` is said to be relocatable if
1. all `include()` paths are relocatable (they start with `@depot`),
2. all `include_dependency()` paths are relocatable (they start with
`@depot` and `track_content=true` was used to include them).

(cherry picked from commit e9d25ca09382b0f67a4c7770cba08bff3db3cb38)

fatteneder committed:
Avoid repeated precompilation when loading from non-relocatable cachefiles (#53905)

Fixes
https://github.com/JuliaLang/julia/issues/53859#issuecomment-2027352004,
which was actually fixed before in
https://github.com/JuliaLang/julia/pull/52346, but
https://github.com/JuliaLang/julia/pull/52750 undid that fix.

This PR does not directly address #53859, because I can not reproduce it
atm.

---

The `@depot` resolution logic for `include()` files is adjusted as
follows:

1. (new behavior) If the cache is not relocatable because of an absolute
path, we ignore that path for the depot search. Recompilation will be
triggered by `stale_cachefile()` if that absolute path does not exist.
Previously this caused any `@depot` tags to be not resolved and so
trigger recompilation.
2. (new behavior) If we can't find a depot for a relocatable path, we
still replace it with the depot we found from other files. Recompilation
will be triggered by `stale_cachefile()` because the resolved path does
not exist.
3. (this behavior is kept) We require that relocatable paths all resolve
to the same depot.
4. (new behavior) We no longer use the first matching depot for
replacement, but instead we explicitly check that all resolve to the
same depot. This has two reasons:
- We want to scan all source files anyways in order to provide logs for
1. and 2. above, so the check is free.
- It is possible that a depot might be missing source files. Assume that
we have two depots on `DEPOT_PATH`, `depot_complete` and
`depot_incomplete`.
If `DEPOT_PATH=["depot_complete","depot_incomplete"]` then no
recompilation shall happen, because `depot_complete` will be picked.
If `DEPOT_PATH=["depot_incomplete","depot_complete"]` we trigger
recompilation and hopefully a meaningful error about missing files is
thrown. If we were to just select the first depot we find, then whether
recompilation happens would depend on whether the first relocatable file
resolves to `depot_complete` or `depot_incomplete`.

(cherry picked from commit d8d384278d6a706e0dc4602bdae1ec1a08908394)

IanButterworth committed:
Profile: make heap snapshots viewable in vscode viewer (#53833)

(cherry picked from commit 57bbff63b8e74c78ffe185296eaba5fd781bc254)

jishnub committed:
`LazyString` in `LinearAlgebra.checksquare` error message (#53961)

This reduces dynamic dispatch and makes JET happier.
Testing on v1.11:
```julia
julia> import LinearAlgebra: checksquare

julia> using JET

julia> @report_opt checksquare(rand(2,2), rand(2,2))
═════ 4 possible errors found ═════
┌ checksquare(::Matrix{Float64}, ::Matrix{Float64}) @ LinearAlgebra /cache/build/builder-amdci4-1/julialang/julia-release-1-dot-11/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/LinearAlgebra.jl:307
│┌ string(::String, ::Tuple{Int64, Int64}) @ Base ./strings/io.jl:189
││┌ print_to_string(::String, ::Tuple{Int64, Int64}) @ Base ./strings/io.jl:150
│││┌ _unsafe_take!(io::IOBuffer) @ Base ./iobuffer.jl:494
││││┌ wrap(::Type{Array}, m::MemoryRef{UInt8}, l::Int64) @ Base ./array.jl:3101
│││││ failed to optimize due to recursion: wrap(::Type{Array}, ::MemoryRef{UInt8}, ::Int64)
││││└────────────────────
│││┌ print_to_string(::String, ::Vararg{Any}) @ Base ./strings/io.jl:143
││││ runtime dispatch detected: Base._str_sizehint(%17::Any)::Int64
│││└────────────────────
│││┌ print_to_string(::String, ::Vararg{Any}) @ Base ./strings/io.jl:148
││││ runtime dispatch detected: print(%59::IOBuffer, %97::Any)::Any
│││└────────────────────
│││┌ string(::String, ::Int64, ::String, ::Tuple{Int64}, ::String, ::Int64, ::String, ::Int64, ::String) @ Base ./strings/io.jl:189
││││ failed to optimize due to recursion: string(::String, ::Int64, ::String, ::Tuple{Int64}, ::String, ::Int64, ::String, ::Int64, ::String)
│││└────────────────────

julia> function checksquare(A...) # This PR
                  sizes = Int[]
                  for a in A
                      size(a,1)==size(a,2) || throw(DimensionMismatch(lazy"matrix is not square: dimensions are $(size(a))"))
                      push!(sizes, size(a,1))
                  end
                  return sizes
              end
checksquare (generic function with 2 methods)

julia> @report_opt checksquare(rand(2,2), rand(2,2))
No errors detected

```

(cherry picked from commit d505c8cdafa79387139f6cf3bf3aa43f0118c087)

Zentrik committed:
Use StringMemory instead of StringVector where possible (#53962)

On `1.11.0-alpha2`
Old:
```julia
@benchmark Base.dec($0x1, $0, $false)
BenchmarkTools.Trial: 10000 samples with 994 evaluations.
 Range (min … max):  33.702 ns …   4.242 μs  ┊ GC (min … max):  0.00% … 97.61%
 Time  (median):     37.626 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   45.787 ns ± 147.794 ns  ┊ GC (mean ± σ):  14.53% ±  4.47%

    ▄▅▆▇█▇▇▅▃▃▂▂▂▁    ▁▂▁▁▁             ▁▁   ▁                 ▂
  ▄▇███████████████▇▇██████▇█▆▆▄▄▃▄▅▄▆▇████████▆▅▅▇▆▅▆▄▄▅▄▄▄▁▅ █
  33.7 ns       Histogram: log(frequency) by time      67.5 ns <

 Memory estimate: 88 bytes, allocs estimate: 3.
```
New:
```julia
BenchmarkTools.Trial: 10000 samples with 995 evaluations.
 Range (min … max):  27.538 ns …   3.397 μs  ┊ GC (min … max):  0.00% … 97.86%
 Time  (median):     30.151 ns               ┊ GC (median):     0.00%
 Time  (mean ± σ):   34.547 ns ± 105.101 ns  ┊ GC (mean ± σ):  10.37% ±  3.39%

       ▁ █▆▃  ▁
  ▂▂▃▃▅█████▆████▆▄▄▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂ ▃
  27.5 ns         Histogram: frequency by time         43.8 ns <

 Memory estimate: 56 bytes, allocs estimate: 2.
```

Fixes #53950, actually now even faster than `1.10.2`.

It doesn't look like the length is ever changed and we don't return
these `StringMemory`s so this change should be fine.

(cherry picked from commit 0e59c9eea225ed186671de1e5fd0c249b1495c1d)

nsajko committed:
profile: doc: update the `Allocs.@profile` doc string (#53825)

(cherry picked from commit d963a3469d742c633bc4d6f0279dd55f19e6c353)

LilithHafner committed:
Make reshape and view on Memory produce Arrays and delete wrap (#53896)

- Make reshape and view with one based indexing on Memory produce Arrays
- delete wrap

Implements
https://github.com/JuliaLang/julia/issues/53552#issuecomment-2024288283

---------

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
(cherry picked from commit 273d91e0dab8ff08d427eaebd58627b2bbd4d807)

jishnub committed:
`LazyString` in `DimensionMismatch` error messages in broadcasting (#53975)

This reduces dynamic dispatch, and makes JET happier. Something similar
is already used in line 523.

(cherry picked from commit f7c7410b8f7d526dad499b87896f92fb39b4e56c)

LilithHafner committed:
Test and fix non-int-length bug in `view(::Memory, ::Union{UnitRange, Base.OneTo})` (#53991)

We assumed, falsely, that `length(inds) isa Int`. The length must be
convertible to an `Int` or we throw, but that conversion may need to be
explicitly performed.

Fixes #53990

CC @oscardssmith @vtjnash @odow

(cherry picked from commit e4f2124e08a14fdfa2c0e569ceb4412b2f666868)
